src/data/list/rotate.lean

```diff
@@ -102,7 +111,233 @@ lemma prod_rotate_eq_one_of_prod_eq_one [group α] : ∀ {l : list α} (hl : l.p
+  induction n with n hn generalizing l,
+  { simp },
+  { cases l with hd tl,
+    { simp },
+    { rw [rotate_cons_succ, nat.succ_eq_add_one, ←rotate_rotate, hn],
+      simp } }
+end
+
+section is_rotated
+
+variables (l l' : list α)
+
+/-- `is_rotated l₁ l₂` or `l₁ ~r l₂` asserts that `l₁` and `l₂` are cyclic permutations
+  of each other. This is defined by claiming that `∃ n, l.rotate n = l'`. -/
+def is_rotated : Prop := ∃ n, l.rotate n = l'
+
+infixr ` ~r `:1000 := is_rotated
+
+variables {l l'}
+
+@[refl] lemma is_rotated.refl (l : list α) : l ~r l :=
+⟨0, by simp⟩
+
+@[symm] lemma is_rotated.symm (h : l ~r l') : l' ~r l :=
+begin
+  obtain ⟨n, rfl⟩ := h,
+  cases l with hd tl,
+  { simp },
+  { use (hd :: tl).length * n - n,
+    rw [rotate_rotate, nat.add_sub_cancel', rotate_length_mul],
+    exact nat.le_mul_of_pos_left (by simp) }
+end
+
+lemma is_rotated_comm : l ~r l' ↔ l' ~r l :=
+⟨is_rotated.symm, is_rotated.symm⟩
+
+@[simp] protected lemma is_rotated.forall (l : list α) (n : ℕ) : l.rotate n ~r l :=
+is_rotated.symm ⟨n, rfl⟩
+
+@[trans] lemma is_rotated.trans {l'' : list α} (h : l ~r l') (h' : l' ~r l'') :
+  l ~r l'' :=
+begin
+  obtain ⟨n, rfl⟩ := h,
+  obtain ⟨m, rfl⟩ := h',
+  rw rotate_rotate,
+  use (n + m)
+end
+
+lemma is_rotated.eqv : equivalence (@is_rotated α) :=
+mk_equivalence _ is_rotated.refl (λ _ _, is_rotated.symm) (λ _ _ _, is_rotated.trans)
+
+/-- The relation `list.is_rotated l l'` forms a `setoid` of cycles. -/
+def is_rotated.setoid (α : Type*) : setoid (list α) :=
+{ r := is_rotated, iseqv := is_rotated.eqv }
+
+lemma is_rotated.perm (h : l ~r l') : l ~ l' :=
+exists.elim h (λ _ hl, hl ▸ (rotate_perm _ _).symm)
+
+lemma is_rotated.nodup_iff (h : l ~r l') : nodup l ↔ nodup l' :=
+h.perm.nodup_iff
+
+lemma is_rotated.mem_iff (h : l ~r l') {a : α} : a ∈ l ↔ a ∈ l' :=
+h.perm.mem_iff
+
+@[simp] lemma is_rotated_nil_iff : l ~r [] ↔ l = [] :=
+⟨λ ⟨n, hn⟩, by simpa using hn, λ h, h ▸ by refl⟩
+
+@[simp] lemma is_rotated_nil_iff' : [] ~r l ↔ l = [] :=
+by rw [is_rotated_comm, is_rotated_nil_iff]
+
+lemma is_rotated_concat (hd : α) (tl : list α) :
+  (tl ++ [hd]) ~r (hd :: tl) :=
+is_rotated.symm ⟨1, by simp⟩
+
+lemma is_rotated.reverse (h : l ~r l') : l.reverse ~r l'.reverse :=
+begin
+  obtain ⟨n, rfl⟩ := h,
+  exact ⟨_, (reverse_rotate _ _).symm⟩
+end
+
+lemma is_rotated_reverse_comm_iff :
+  l.reverse ~r l' ↔ l ~r l'.reverse :=
+begin
+  split;
+  { intro h,
+    simpa using h.reverse }
+end
+
+@[simp] lemma is_rotated_reverse_iff :
+  l.reverse ~r l'.reverse ↔ l ~r l' :=
+by simp [is_rotated_reverse_comm_iff]
+
+lemma is_rotated_iff_mod : l ~r l' ↔ ∃ n ≤ l.length, l.rotate n = l' :=
+begin
+  refine ⟨λ h, _, λ ⟨n, _, h⟩, ⟨n, h⟩⟩,
+  obtain ⟨n, rfl⟩ := h,
+  cases l with hd tl,
+  { simp },
+  { refine ⟨n % (hd :: tl).length, _, rotate_mod _ _⟩,
+    refine (nat.mod_lt _ _).le,
+    simp }
+end
+
+lemma is_rotated_iff_mem_map_range : l ~r l' ↔ l' ∈ (list.range (l.length + 1)).map l.rotate :=
```

Comment:
Isn't the stronger statement true, 
```suggestion
lemma is_rotated_iff_mem_map_range : l ~r l' ↔ l' ∈ (list.range l.length).map l.rotate :=
```

Reply:
No, because if `l = []` then `list.range nil.length = list.range 0 = []` and `[] \notin []`.

Reply:
Ah, not when `l = []`. This is all quite annoying. Is it worth a variant:

```suggestion
lemma is_rotated_iff_mem_map_range (hl : l = []) : l ~r l' ↔ l' ∈ (list.range l.length).map l.rotate :=
```

Reply:
I think you mean `\ne` in your statement. I think that's an annoying hypothesis to have to create. I hope to use this iff statement to be able to provide a computable way of escaping the `cycle` quotient I am constructing, when working over a type with an order -- sort the `(list.range ...).map l.rotate` by lexicographic order and pick the first one. Which works also for the nil list too.